In [77]:
import pandas as pd
path = "C:/Users/super/OneDrive - The University of Tokyo/Portfolio Research work with Korbtham/Project_Bioisostere_Prediction/TryProject1/Dataset/ci0c00290_si_002.txt"
data = pd.read_csv(path,sep='  ',header=None)

<ipython-input-77-2ebd0f57911b>:3: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  data = pd.read_csv(path,sep='  ',header=None)


In [78]:
data

0
0               [R]C\t*
1          [R]CC\t14613
2          [R]Cl\t11518
3           [R]OC\t8873
4            [R]F\t8034
...                 ...
2095  [R]NC(C)(C)C\t188
2096          [R]O\t187
2097   [R]N1CCCCC1\t174
2098         [R]CC\t162
2099   [R]NC1CCCC1\t160

[2100 rows x 1 columns]

In [79]:
data['1'] = [0 for i in range(0,2100)]
data

0  1
0               [R]C\t*  0
1          [R]CC\t14613  0
2          [R]Cl\t11518  0
3           [R]OC\t8873  0
4            [R]F\t8034  0
...                 ... ..
2095  [R]NC(C)(C)C\t188  0
2096          [R]O\t187  0
2097   [R]N1CCCCC1\t174  0
2098         [R]CC\t162  0
2099   [R]NC1CCCC1\t160  0

[2100 rows x 2 columns]

In [80]:
n,m = data.shape

In [81]:
for i in range(n):
    lis = data.iloc[i,0].split('\t')
    data.iloc[i,0] = lis[0]
    data.iloc[i,1] = lis[1]

In [82]:
#remove[R]
for i in range(n):
    data.iloc[i,0] = data.iloc[i,0].replace('[R]','')

In [83]:
data

0      1
0             C      *
1            CC  14613
2            Cl  11518
3            OC   8873
4             F   8034
...         ...    ...
2095  NC(C)(C)C    188
2096          O    187
2097   N1CCCCC1    174
2098         CC    162
2099   NC1CCCC1    160

[2100 rows x 2 columns]

In [84]:
# Attach the row of substitutes data
n,m = data.shape
lis_sub = []
for i in range(n):
    if data.iloc[i,1] == '*':
        lis_sub.append(data.iloc[i,0])

In [85]:
lis_sub

['C',
 'O',
 'F',
 'N',
 'c1ccccc1',
 'Cl',
 'OC',
 'C(O)=O',
 'CC',
 'C(C)C',
 'Cc1ccccc1',
 'C(F)(F)F',
 'C(N)=O',
 'C#N',
 'CC(C)C',
 'c1ccc(Cl)cc1',
 'c1ccc(F)cc1',
 'C(C)(C)C',
 'CO',
 'C(C)=O',
 'Br',
 'N=C(N)N',
 'CC(O)=O',
 'C(N)=N',
 'CCC',
 'CCCN=C(N)N',
 'N(C)C',
 'c1ccc(OC)cc1',
 'c1ccc(O)cc1',
 'OCC',
 'C1CC1',
 'C1CCCCC1',
 'N1CCOCC1',
 'S(C)(=O)=O',
 'CCCC',
 'N(=O)=O',
 'NC',
 'CCC(O)=O',
 'C(=O)OC',
 'CCCCN',
 'c1c[nH]c2ccccc12',
 'NC(C)=O',
 'NO',
 'S(N)(=O)=O',
 'c1cccnc1',
 'C(=O)NO',
 'Cc1ccc(O)cc1',
 'c1ccccn1',
 'c1ccc(C)cc1',
 'C(C)CC',
 'Cc1c[nH]c2ccccc12',
 'OCc1ccccc1',
 'N1CCCCC1',
 'c1ccncc1',
 'C(C)O',
 'c1cccc(Cl)c1',
 'N1CCN(C)CC1',
 'CC(N)=O',
 'c1ccccc1Cl',
 'c1cnc[nH]1',
 'N1CCCC1',
 'C(=O)OCC',
 'CCc1ccccc1',
 'c1ccccc1OC',
 'c2ccc1ccccc1c2',
 'SC',
 'c1ccc(C(F)(F)F)cc1',
 'Oc1ccccc1',
 'c1ccc(Cl)c(Cl)c1',
 'c1cccc(OC)c1',
 'C1CCCC1',
 'Cc1cnc[nH]1',
 'C(=O)c1ccccc1',
 'CCC(N)=O',
 'c1cccc(F)c1',
 'OC(C)(C)C',
 'C(=O)OC(C)(C)C',
 'c1cccs1',
 'c1ccc(C

Then we will pefrom fragment descriptors for each substitues

In [86]:
data_encod = data.copy()

In [87]:
data_encod.columns

Index([0, '1'], dtype='object')

#https://www.blopig.com/blog/2022/06/how-to-turn-a-molecule-into-a-vector-of-physicochemical-descriptors-using-rdkit/

In [88]:
#We will perform the molecular desciptors from smiles strucutre
from rdkit import Chem
from rdkit.ML.Descriptors.MoleculeDescriptors import MolecularDescriptorCalculator

In [89]:
smiles_string = 'CN'

In [90]:
mol = Chem.MolFromSmiles(smiles_string)

In [91]:
chosen_descriptors = ['BalabanJ', 'BertzCT', 'Chi0', 'Chi0n', 'Chi0v', 'Chi1', 'Chi1n', 'Chi1v', 'Chi2n', 'Chi2v', 'Chi3n', 'Chi3v', 'Chi4n', 'Chi4v', 'EState_VSA1', 'EState_VSA10', 'EState_VSA11', 'EState_VSA2', 'EState_VSA3', 'EState_VSA4', 'EState_VSA5', 'EState_VSA6', 'EState_VSA7', 'EState_VSA8', 'EState_VSA9', 'ExactMolWt', 'FpDensityMorgan1', 'FpDensityMorgan2', 'FpDensityMorgan3', 'FractionCSP3', 'HallKierAlpha', 'HeavyAtomCount', 'HeavyAtomMolWt', 'Ipc', 'Kappa1', 'Kappa2', 'Kappa3', 'LabuteASA', 'MaxAbsEStateIndex', 'MaxAbsPartialCharge', 'MaxEStateIndex', 'MaxPartialCharge', 'MinAbsEStateIndex', 'MinAbsPartialCharge', 'MinEStateIndex', 'MinPartialCharge', 'MolLogP', 'MolMR', 'MolWt', 'NHOHCount', 'NOCount', 'NumAliphaticCarbocycles', 'NumAliphaticHeterocycles', 'NumAliphaticRings', 'NumAromaticCarbocycles', 'NumAromaticHeterocycles', 'NumAromaticRings', 'NumHAcceptors', 'NumHDonors', 'NumHeteroatoms', 'NumRadicalElectrons', 'NumRotatableBonds', 'NumSaturatedCarbocycles', 'NumSaturatedHeterocycles', 'NumSaturatedRings', 'NumValenceElectrons', 'PEOE_VSA1', 'PEOE_VSA10', 'PEOE_VSA11', 'PEOE_VSA12', 'PEOE_VSA13', 'PEOE_VSA14', 'PEOE_VSA2', 'PEOE_VSA3', 'PEOE_VSA4', 'PEOE_VSA5', 'PEOE_VSA6', 'PEOE_VSA7', 'PEOE_VSA8', 'PEOE_VSA9', 'RingCount', 'SMR_VSA1', 'SMR_VSA10', 'SMR_VSA2', 'SMR_VSA3', 'SMR_VSA4', 'SMR_VSA5', 'SMR_VSA6', 'SMR_VSA7', 'SMR_VSA8', 'SMR_VSA9', 'SlogP_VSA1', 'SlogP_VSA10', 'SlogP_VSA11', 'SlogP_VSA12', 'SlogP_VSA2', 'SlogP_VSA3', 'SlogP_VSA4', 'SlogP_VSA5', 'SlogP_VSA6', 'SlogP_VSA7', 'SlogP_VSA8', 'SlogP_VSA9', 'TPSA', 'VSA_EState1', 'VSA_EState10', 'VSA_EState2', 'VSA_EState3', 'VSA_EState4', 'VSA_EState5', 'VSA_EState6', 'VSA_EState7', 'VSA_EState8', 'VSA_EState9', 'fr_Al_COO', 'fr_Al_OH', 'fr_Al_OH_noTert', 'fr_ArN', 'fr_Ar_COO', 'fr_Ar_N', 'fr_Ar_NH', 'fr_Ar_OH', 'fr_COO', 'fr_COO2', 'fr_C_O', 'fr_C_O_noCOO', 'fr_C_S', 'fr_HOCCN', 'fr_Imine', 'fr_NH0', 'fr_NH1', 'fr_NH2', 'fr_N_O', 'fr_Ndealkylation1', 'fr_Ndealkylation2', 'fr_Nhpyrrole', 'fr_SH', 'fr_aldehyde', 'fr_alkyl_carbamate', 'fr_alkyl_halide', 'fr_allylic_oxid', 'fr_amide', 'fr_amidine', 'fr_aniline', 'fr_aryl_methyl', 'fr_azide', 'fr_azo', 'fr_barbitur', 'fr_benzene', 'fr_benzodiazepine', 'fr_bicyclic', 'fr_diazo', 'fr_dihydropyridine', 'fr_epoxide', 'fr_ester', 'fr_ether', 'fr_furan', 'fr_guanido', 'fr_halogen', 'fr_hdrzine', 'fr_hdrzone', 'fr_imidazole', 'fr_imide', 'fr_isocyan', 'fr_isothiocyan', 'fr_ketone', 'fr_ketone_Topliss', 'fr_lactam', 'fr_lactone', 'fr_methoxy', 'fr_morpholine', 'fr_nitrile', 'fr_nitro', 'fr_nitro_arom', 'fr_nitro_arom_nonortho', 'fr_nitroso', 'fr_oxazole', 'fr_oxime', 'fr_para_hydroxylation', 'fr_phenol', 'fr_phenol_noOrthoHbond', 'fr_phos_acid', 'fr_phos_ester', 'fr_piperdine', 'fr_piperzine', 'fr_priamide', 'fr_prisulfonamd', 'fr_pyridine', 'fr_quatN', 'fr_sulfide', 'fr_sulfonamd', 'fr_sulfone', 'fr_term_acetylene', 'fr_tetrazole', 'fr_thiazole', 'fr_thiocyan', 'fr_thiophene', 'fr_unbrch_alkane', 'fr_urea', 'qed']
# create molecular descriptor calculator
mol_descriptor_calculator = MolecularDescriptorCalculator(chosen_descriptors)
# use molecular descriptor calculator on RDKit mol object
list_of_descriptor_vals = list(mol_descriptor_calculator.CalcDescriptors(mol))


In [92]:
# We will generate the dataset containing molecular desciptors and smile string 
# rename the column of data_encod
data_encod.rename(columns = {0:'SMILES','1':'Count'}, inplace = True)

In [ ]:
for des in chosen_descriptors:
    data_encod[des] = [0 for i in range(0,2100)]

In [ ]:
for i in range(0,2100):
    smile_string = data_encod.iloc[i,0]
    mol = Chem.MolFromSmiles(smiles_string)
    mol_descriptor_calculator = MolecularDescriptorCalculator(chosen_descriptors)
    list_of_descriptor_vals = list(mol_descriptor_calculator.CalcDescriptors(mol)
    data_encod.iloc[i,2:202] = list_of_descriptor_vals
    

In [95]:
for i in range(0,2100):
    smiles_string = data_encod.iloc[i,0]
    # convert SMILES string to RDKit mol object
    mol = Chem.MolFromSmiles(smiles_string)
    # choose 200 molecular descriptors
    chosen_descriptors = ['BalabanJ', 'BertzCT', 'Chi0', 'Chi0n', 'Chi0v', 'Chi1', 'Chi1n', 'Chi1v', 'Chi2n', 'Chi2v', 'Chi3n', 'Chi3v', 'Chi4n', 'Chi4v', 'EState_VSA1', 'EState_VSA10', 'EState_VSA11', 'EState_VSA2', 'EState_VSA3', 'EState_VSA4', 'EState_VSA5', 'EState_VSA6', 'EState_VSA7', 'EState_VSA8', 'EState_VSA9', 'ExactMolWt', 'FpDensityMorgan1', 'FpDensityMorgan2', 'FpDensityMorgan3', 'FractionCSP3', 'HallKierAlpha', 'HeavyAtomCount', 'HeavyAtomMolWt', 'Ipc', 'Kappa1', 'Kappa2', 'Kappa3', 'LabuteASA', 'MaxAbsEStateIndex', 'MaxAbsPartialCharge', 'MaxEStateIndex', 'MaxPartialCharge', 'MinAbsEStateIndex', 'MinAbsPartialCharge', 'MinEStateIndex', 'MinPartialCharge', 'MolLogP', 'MolMR', 'MolWt', 'NHOHCount', 'NOCount', 'NumAliphaticCarbocycles', 'NumAliphaticHeterocycles', 'NumAliphaticRings', 'NumAromaticCarbocycles', 'NumAromaticHeterocycles', 'NumAromaticRings', 'NumHAcceptors', 'NumHDonors', 'NumHeteroatoms', 'NumRadicalElectrons', 'NumRotatableBonds', 'NumSaturatedCarbocycles', 'NumSaturatedHeterocycles', 'NumSaturatedRings', 'NumValenceElectrons', 'PEOE_VSA1', 'PEOE_VSA10', 'PEOE_VSA11', 'PEOE_VSA12', 'PEOE_VSA13', 'PEOE_VSA14', 'PEOE_VSA2', 'PEOE_VSA3', 'PEOE_VSA4', 'PEOE_VSA5', 'PEOE_VSA6', 'PEOE_VSA7', 'PEOE_VSA8', 'PEOE_VSA9', 'RingCount', 'SMR_VSA1', 'SMR_VSA10', 'SMR_VSA2', 'SMR_VSA3', 'SMR_VSA4', 'SMR_VSA5', 'SMR_VSA6', 'SMR_VSA7', 'SMR_VSA8', 'SMR_VSA9', 'SlogP_VSA1', 'SlogP_VSA10', 'SlogP_VSA11', 'SlogP_VSA12', 'SlogP_VSA2', 'SlogP_VSA3', 'SlogP_VSA4', 'SlogP_VSA5', 'SlogP_VSA6', 'SlogP_VSA7', 'SlogP_VSA8', 'SlogP_VSA9', 'TPSA', 'VSA_EState1', 'VSA_EState10', 'VSA_EState2', 'VSA_EState3', 'VSA_EState4', 'VSA_EState5', 'VSA_EState6', 'VSA_EState7', 'VSA_EState8', 'VSA_EState9', 'fr_Al_COO', 'fr_Al_OH', 'fr_Al_OH_noTert', 'fr_ArN', 'fr_Ar_COO', 'fr_Ar_N', 'fr_Ar_NH', 'fr_Ar_OH', 'fr_COO', 'fr_COO2', 'fr_C_O', 'fr_C_O_noCOO', 'fr_C_S', 'fr_HOCCN', 'fr_Imine', 'fr_NH0', 'fr_NH1', 'fr_NH2', 'fr_N_O', 'fr_Ndealkylation1', 'fr_Ndealkylation2', 'fr_Nhpyrrole', 'fr_SH', 'fr_aldehyde', 'fr_alkyl_carbamate', 'fr_alkyl_halide', 'fr_allylic_oxid', 'fr_amide', 'fr_amidine', 'fr_aniline', 'fr_aryl_methyl', 'fr_azide', 'fr_azo', 'fr_barbitur', 'fr_benzene', 'fr_benzodiazepine', 'fr_bicyclic', 'fr_diazo', 'fr_dihydropyridine', 'fr_epoxide', 'fr_ester', 'fr_ether', 'fr_furan', 'fr_guanido', 'fr_halogen', 'fr_hdrzine', 'fr_hdrzone', 'fr_imidazole', 'fr_imide', 'fr_isocyan', 'fr_isothiocyan', 'fr_ketone', 'fr_ketone_Topliss', 'fr_lactam', 'fr_lactone', 'fr_methoxy', 'fr_morpholine', 'fr_nitrile', 'fr_nitro', 'fr_nitro_arom', 'fr_nitro_arom_nonortho', 'fr_nitroso', 'fr_oxazole', 'fr_oxime', 'fr_para_hydroxylation', 'fr_phenol', 'fr_phenol_noOrthoHbond', 'fr_phos_acid', 'fr_phos_ester', 'fr_piperdine', 'fr_piperzine', 'fr_priamide', 'fr_prisulfonamd', 'fr_pyridine', 'fr_quatN', 'fr_sulfide', 'fr_sulfonamd', 'fr_sulfone', 'fr_term_acetylene', 'fr_tetrazole', 'fr_thiazole', 'fr_thiocyan', 'fr_thiophene', 'fr_unbrch_alkane', 'fr_urea', 'qed']
    # create molecular descriptor calculator
    mol_descriptor_calculator = MolecularDescriptorCalculator(chosen_descriptors)
    # use molecular descriptor calculator on RDKit mol object
    list_of_descriptor_vals = list(mol_descriptor_calculator.CalcDescriptors(mol))
    data_encod.iloc[i,2:202] = list_of_descriptor_vals

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 463, in BalabanJ
    dMat = mol._balabanMat
AttributeError: 'NoneType' object has no attribute '_balabanMat'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\ML\Descriptors\MoleculeDescriptors.py", line 88, in CalcDescriptors
    res[i] = fn(mol)
  File "C:\Users\super\anaconda3\lib\site-packages\rdkit\Chem\GraphDescriptors.py", line 466, in BalabanJ
    dMat = Chem.GetDistanceMatrix(mol, useBO=1, useAtomWts=0, force=0, prefix="Balaban")
Boost.Python.ArgumentError: Python argument types in
    rdkit.Chem.rdmolops.GetDistanceMatrix(NoneType)
did not match C++ signature:
    GetDistanceMatrix(class RDKit::ROMol {lvalue} mol, bool useBO=False, bool useAtomWts=False, bool force=False, char const * __ptr64 prefix='')
Traceback (most recent call last):
  File

In [96]:
data_encod

SMILES  Count  BalabanJ    BertzCT      Chi0     Chi0n     Chi0v  \
0             C      *  0.000000   0.000000  0.000000  0.000000  0.000000   
1            CC  14613  1.000000   0.000000  2.000000  2.000000  2.000000   
2            Cl  11518  0.000000   0.000000  0.000000  0.408248  1.224745   
3            OC   8873  1.000000   2.000000  2.000000  1.447214  1.447214   
4             F   8034  0.000000   0.000000  0.000000  0.408248  0.408248   
...         ...    ...       ...        ...       ...       ...       ...   
2095  NC(C)(C)C    188  3.023716  19.119415  4.500000  4.077350  4.077350   
2096          O    187  0.000000   0.000000  0.000000  0.500000  0.500000   
2097   N1CCCCC1    174  2.000000  19.409910  4.242641  4.035534  4.035534   
2098         CC    162  1.000000   0.000000  2.000000  2.000000  2.000000   
2099   NC1CCCC1    160  2.184106  37.203103  4.405777  3.983128  3.983128   

          Chi1     Chi1n     Chi1v  ...  fr_sulfonamd  fr_sulfone  \
0     0.000000  0.000000  0.000000  ...             0           0   
1     1.000000  1.000000  1.000000  ...             0           0   
2     0.000000  0.000000  0.000000  ...             0           0   
3     1.000000  0.447214  0.447214  ...             0           0   
4     0.000000  0.000000  0.000000  ...             0           0   
...        ...       ...       ...  ...           ...         ...   
2095  2.000000  1.788675  1.788675  ...             0           0   
2096  0.000000  0.000000  0.000000  ...             0           0   
2097  3.000000  2.707107  2.707107  ...             0           0   
2098  1.000000  1.000000  1.000000  ...             0           0   
2099  2.893847  2.649830  2.649830  ...             0           0   

      fr_term_acetylene  fr_tetrazole  fr_thiazole  fr_thiocyan  fr_thiophene  \
0                     0             0            0            0             0   
1                     0             0            0            0             0   
2                     0             0            0            0             0   
3                     0             0            0            0             0   
4                     0             0            0            0             0   
...                 ...           ...          ...          ...           ...   
2095                  0             0            0            0             0   
2096                  0             0            0            0             0   
2097                  0             0            0            0             0   
2098                  0             0            0            0             0   
2099                  0             0            0            0             0   

      fr_unbrch_alkane  fr_urea       qed  
0                    0        0  0.359785  
1                    0        0  0.372786  
2                    0        0  0.365164  
3                    0        0  0.385284  
4                    0        0  0.350795  
...                ...      ...       ...  
2095                 0        0  0.449470  
2096                 0        0  0.327748  
2097                 0        0  0.457636  
2098                 0        0  0.372786  
2099                 0        0  0.464132  

[2100 rows x 202 columns]

In [97]:
# Save encoded data to csv file
data_encod.to_csv('preprocess_descriptors_data.csv')

Since there are many descriptors(around 200 descriptors), so we need to perform dimension reduction used principal component analysis 

In [98]:
# First, we need to perform standardized before principal analysis
data_encod = data_encod[chosen_descriptors]
from sklearn.preprocessing import StandardScaler
scalar = StandardScaler()
data_scaled = pd.DataFrame(scalar.fit_transform(data_encod), columns=data_encod.columns)
data_scaled

BalabanJ   BertzCT      Chi0     Chi0n     Chi0v      Chi1     Chi1n  \
0     0.062989 -0.596359  0.037163  0.049214  0.046667  0.060303  0.072426   
1     0.078395 -0.596359  0.067886  0.079978  0.077424  0.075703  0.087846   
2     0.062989 -0.596359  0.037163  0.055494  0.065502  0.060303  0.072426   
3     0.078395 -0.577684  0.067886  0.071475  0.068923  0.075703  0.079322   
4     0.062989 -0.596359  0.037163  0.055494  0.052946  0.060303  0.072426   
...        ...       ...       ...       ...       ...       ...       ...   
2095  0.109570 -0.417832  0.106289  0.111932  0.109370  0.091102  0.100007   
2096  0.062989 -0.596359  0.037163  0.056905  0.054357  0.060303  0.072426   
2097  0.093800 -0.415119  0.102336  0.111289  0.108727  0.106502  0.114168   
2098  0.078395 -0.596359  0.067886  0.079978  0.077424  0.075703  0.087846   
2099  0.096636 -0.248976  0.104842  0.110483  0.107921  0.104867  0.113285   

         Chi1v     Chi2n     Chi2v  ...  fr_sulfonamd  fr_sulfone  \
0     0.070679  0.082087  0.080540  ...      0.097902    0.098051   
1     0.086096  0.082087  0.080540  ...      0.097902    0.098051   
2     0.070679  0.082087  0.080540  ...      0.097902    0.098051   
3     0.077574  0.082087  0.080540  ...      0.097902    0.098051   
4     0.070679  0.082087  0.080540  ...      0.097902    0.098051   
...        ...       ...       ...  ...           ...         ...   
2095  0.098255  0.118606  0.117053  ...      0.097902    0.098051   
2096  0.070679  0.082087  0.080540  ...      0.097902    0.098051   
2097  0.112414  0.110034  0.108482  ...      0.097902    0.098051   
2098  0.086096  0.082087  0.080540  ...      0.097902    0.098051   
2099  0.111531  0.113644  0.112091  ...      0.097902    0.098051   

      fr_term_acetylene  fr_tetrazole  fr_thiazole  fr_thiocyan  fr_thiophene  \
0              0.098051      0.098021     0.098036     0.098058      0.097909   
1              0.098051      0.098021     0.098036     0.098058      0.097909   
2              0.098051      0.098021     0.098036     0.098058      0.097909   
3              0.098051      0.098021     0.098036     0.098058      0.097909   
4              0.098051      0.098021     0.098036     0.098058      0.097909   
...                 ...           ...          ...          ...           ...   
2095           0.098051      0.098021     0.098036     0.098058      0.097909   
2096           0.098051      0.098021     0.098036     0.098058      0.097909   
2097           0.098051      0.098021     0.098036     0.098058      0.097909   
2098           0.098051      0.098021     0.098036     0.098058      0.097909   
2099           0.098051      0.098021     0.098036     0.098058      0.097909   

      fr_unbrch_alkane   fr_urea       qed  
0             0.097946  0.098028  0.096813  
1             0.097946  0.098028  0.097014  
2             0.097946  0.098028  0.096896  
3             0.097946  0.098028  0.097207  
4             0.097946  0.098028  0.096674  
...                ...       ...       ...  
2095          0.097946  0.098028  0.098198  
2096          0.097946  0.098028  0.096318  
2097          0.097946  0.098028  0.098324  
2098          0.097946  0.098028  0.097014  
2099          0.097946  0.098028  0.098425  

[2100 rows x 200 columns]

In [99]:
from sklearn.decomposition import PCA
pca = PCA()
data_pca = pd.DataFrame(pca.fit_transform(data_scaled))
data_pca


0         1          2         3         4         5         6    \
0    -0.925417 -1.440281  -0.207667  0.553537 -0.356964 -0.148077 -0.057758   
1    -1.029991 -1.239298  -0.184170  0.373666 -0.343035  0.056124 -0.004112   
2    -1.010439 -1.191738  -0.159094  0.300800 -0.135012 -0.101390  0.230339   
3    -1.066468 -1.234921  -0.230248  0.388332  0.069380 -0.253428  0.063509   
4    -0.902455 -1.390604  -0.324846  0.511796 -0.258633 -0.257729  0.051895   
...        ...       ...        ...       ...       ...       ...       ...   
2095 -1.521148 -1.814591  18.641795  0.375008 -0.514494 -0.043506  0.045231   
2096 -0.922884 -1.427348  -0.404601  0.564180 -0.002328 -0.441646 -0.060426   
2097 -1.409601 -0.299013  -0.081211 -0.029651  0.305961  0.587561 -0.066094   
2098 -1.029991 -1.239298  -0.184170  0.373666 -0.343035  0.056124 -0.004112   
2099 -1.436211 -0.224877  -0.073469 -0.051014  0.340139  0.522850 -0.281987   

           7         8         9    ...           190           191  \
0     0.163211  0.049007 -0.040079  ...  2.148296e-17  1.122230e-18   
1     0.021482 -0.002859  0.071902  ... -1.604568e-16  5.870643e-16   
2     0.250945  0.235519 -0.148127  ... -2.195981e-16  9.268949e-17   
3     0.023479 -0.090876 -0.039300  ... -1.688551e-15 -6.317666e-16   
4     0.252168  0.063259 -0.052793  ... -3.792362e-16 -7.931659e-16   
...        ...       ...       ...  ...           ...           ...   
2095 -0.161690 -0.000029  0.010396  ... -2.068504e-15  8.729520e-16   
2096  0.007255  0.116220  0.094804  ...  2.586332e-15  2.185755e-15   
2097 -0.087761 -0.229037 -0.192202  ...  8.296721e-16 -5.464038e-16   
2098  0.021482 -0.002859  0.071902  ... -1.706062e-17  3.483839e-16   
2099 -0.166047 -0.036942  0.078432  ... -2.735860e-15 -1.019113e-15   

               192           193           194           195           196  \
0    -1.022050e-17  1.306290e-17 -9.896932e-18 -1.675533e-17  1.050802e-17   
1     5.115248e-16 -2.303684e-16  2.125195e-16 -7.238308e-17 -7.030860e-17   
2     1.033637e-15 -2.781405e-17  1.637940e-16  3.074990e-16  4.783702e-17   
3     5.984661e-16 -2.388567e-16  6.980693e-16  9.170374e-16 -7.725410e-16   
4    -2.034002e-16  2.840635e-17 -6.443781e-17 -1.260725e-15  1.125647e-15   
...            ...           ...           ...           ...           ...   
2095 -6.060668e-16  3.243862e-16  5.100212e-16 -2.964358e-16 -1.833581e-15   
2096 -1.767267e-16  1.318750e-15 -4.359687e-16 -6.093062e-16  6.630155e-16   
2097 -1.000675e-15  1.359019e-15 -4.187381e-16  6.129437e-16 -6.380391e-16   
2098 -1.748335e-16  4.045342e-16  2.383338e-16  7.422019e-17 -5.520279e-16   
2099  1.896658e-15 -4.208505e-16  2.156954e-15  2.243512e-15  3.202338e-16   

               197           198           199  
0    -1.729774e-18  1.884399e-17  1.679838e-14  
1     1.802001e-16 -1.359227e-16  9.312064e-15  
2     2.548310e-16 -4.625534e-16 -8.292821e-16  
3     6.015672e-16  4.256678e-16  4.808337e-15  
4    -1.297942e-15  7.594665e-16 -1.285809e-15  
...            ...           ...           ...  
2095  5.670237e-16  4.532752e-15 -1.018030e-15  
2096 -1.375395e-16 -3.102237e-16 -3.939257e-16  
2097  4.384485e-16  6.916876e-16 -3.514216e-16  
2098 -1.478165e-15 -1.058098e-16 -7.543156e-16  
2099  5.795555e-16 -9.784370e-16  1.011957e-16  

[2100 rows x 200 columns]

In [100]:
pca.explained_variance_ratio_

array([9.83497790e-01, 8.03815802e-03, 4.90728669e-03, 1.41849252e-03,
       5.93770387e-04, 4.89539061e-04, 2.10688476e-04, 1.77164047e-04,
       1.09304881e-04, 1.04746204e-04, 7.65001190e-05, 5.92760245e-05,
       4.29218098e-05, 3.94317152e-05, 2.98704542e-05, 2.83491055e-05,
       2.07097801e-05, 1.97384103e-05, 1.61480342e-05, 1.53331022e-05,
       1.22336802e-05, 1.11912407e-05, 1.01557870e-05, 8.54121588e-06,
       7.29968763e-06, 5.53319852e-06, 5.24729241e-06, 4.74478137e-06,
       4.08842982e-06, 3.41532884e-06, 3.33852643e-06, 3.24359770e-06,
       2.67789440e-06, 2.37915660e-06, 2.20546209e-06, 2.07197095e-06,
       1.89245192e-06, 1.70214196e-06, 1.52577648e-06, 1.39072832e-06,
       1.22354175e-06, 1.10044907e-06, 8.72251090e-07, 8.43380529e-07,
       6.88848090e-07, 6.14252408e-07, 5.62739604e-07, 5.14639951e-07,
       4.10742106e-07, 3.38370857e-07, 3.27066685e-07, 2.99396121e-07,
       2.41229809e-07, 2.27983445e-07, 1.83893154e-07, 1.66375434e-07,
      

In [101]:
df_explained = pd.DataFrame(pca.explained_variance_ratio_)

In [102]:
df_explained[0]

0      9.834978e-01
1      8.038158e-03
2      4.907287e-03
3      1.418493e-03
4      5.937704e-04
           ...     
195    9.767634e-33
196    9.767634e-33
197    9.767634e-33
198    9.767634e-33
199    3.270762e-33
Name: 0, Length: 200, dtype: float64

In [103]:
import matplotlib.pyplot as plt
import numpy as np
# calculate cumulative sum of explained variances
tot = sum(df_explained[0])
var_exp = [(i / tot) for i in sorted(df_explained[0], reverse=True)]
cum_var_exp = np.cumsum(var_exp)
var_exp

[0.9834977903886901,
 0.008038158016690385,
 0.004907286691919768,
 0.0014184925238596667,
 0.0005937703874995512,
 0.0004895390611432744,
 0.00021068847558892493,
 0.0001771640465913787,
 0.00010930488068283847,
 0.00010474620354913282,
 7.650011901498861e-05,
 5.927602447273758e-05,
 4.292180980723884e-05,
 3.943171524919659e-05,
 2.9870454194644223e-05,
 2.8349105507067105e-05,
 2.0709780081916422e-05,
 1.9738410328963197e-05,
 1.6148034194468853e-05,
 1.5333102247073618e-05,
 1.2233680167009628e-05,
 1.1191240655522847e-05,
 1.0155786982385724e-05,
 8.541215882293097e-06,
 7.29968763452789e-06,
 5.533198519266753e-06,
 5.247292413511853e-06,
 4.74478137047371e-06,
 4.088429820796095e-06,
 3.4153288410735e-06,
 3.3385264259159174e-06,
 3.2435977038716295e-06,
 2.677894395635275e-06,
 2.3791565973893023e-06,
 2.2054620934799e-06,
 2.0719709494471633e-06,
 1.8924519238833896e-06,
 1.7021419557108637e-06,
 1.5257764775695587e-06,
 1.3907283182131399e-06,
 1.2235417477566682e-06,
 1.100

Based on the PCA result, we see that the only first components capture almost 98.3% of the data. However, we will use two components for this model 

In [104]:
data[['PCA1','PCA2']] = data_pca.iloc[:,0:2]

In [105]:
data

0      1      PCA1      PCA2
0             C      * -0.925417 -1.440281
1            CC  14613 -1.029991 -1.239298
2            Cl  11518 -1.010439 -1.191738
3            OC   8873 -1.066468 -1.234921
4             F   8034 -0.902455 -1.390604
...         ...    ...       ...       ...
2095  NC(C)(C)C    188 -1.521148 -1.814591
2096          O    187 -0.922884 -1.427348
2097   N1CCCCC1    174 -1.409601 -0.299013
2098         CC    162 -1.029991 -1.239298
2099   NC1CCCC1    160 -1.436211 -0.224877

[2100 rows x 4 columns]

In [106]:
# Save as an input X training data
data.to_csv('X_train.csv')

Perform multi-hot encoding on the dataset

1) we will create the list for substitutes data

2) we will create the list that contains all of analogs

In [107]:
data.rename(columns = {0:'SMILES', '1':'Value'}, inplace = True)

In [108]:
# extract the substitutes dataset 
sub = data[data['Value']=='*']
lis_sub = list(sub['SMILES'])

In [109]:
# The list contain all of analog
lis_ana = []
n,m = data.shape
for i in range(n):
    k = data.iloc[i,0]
    if k not in lis_ana:
        lis_ana.append(k)

In [110]:
#Then we will generate multi-hor encoded array 
mult_encod = np.zeros((len(lis_sub),len(lis_ana)))

In [111]:
lis_i = []
for i in range(n):
    if data.iloc[i,1] == '*':
       lis_i.append(i)
    

In [112]:
list(data.iloc[1:22,0])

['CC',
 'Cl',
 'OC',
 'F',
 'C(F)(F)F',
 'C(C)C',
 'c1ccccc1',
 'CCC',
 'Br',
 'N',
 'O',
 'CCCC',
 'C#N',
 'C1CC1',
 'C(C)(C)C',
 'CO',
 'CC(C)C',
 'N(=O)=O',
 'CCO',
 'CC1CC1',
 'O']

In [113]:
dic = {}
for i in range(2100):
    if i%21 ==0:
        print(i)
        key = data.iloc[i,0]
        lis = list(data.iloc[i+1:i+20,0])
        dic[key] = lis

        

0
21
42
63
84
105
126
147
168
189
210
231
252
273
294
315
336
357
378
399
420
441
462
483
504
525
546
567
588
609
630
651
672
693
714
735
756
777
798
819
840
861
882
903
924
945
966
987
1008
1029
1050
1071
1092
1113
1134
1155
1176
1197
1218
1239
1260
1281
1302
1323
1344
1365
1386
1407
1428
1449
1470
1491
1512
1533
1554
1575
1596
1617
1638
1659
1680
1701
1722
1743
1764
1785
1806
1827
1848
1869
1890
1911
1932
1953
1974
1995
2016
2037
2058
2079


In [114]:
print(n)

2100


In [115]:
dic

{'C': ['CC',
  'Cl',
  'OC',
  'F',
  'C(F)(F)F',
  'C(C)C',
  'c1ccccc1',
  'CCC',
  'Br',
  'N',
  'O',
  'CCCC',
  'C#N',
  'C1CC1',
  'C(C)(C)C',
  'CO',
  'CC(C)C',
  'N(=O)=O',
  'CCO'],
 'O': ['OC',
  'N',
  'F',
  'C',
  'OCC',
  'Cl',
  'N(C)C',
  'C(N)=O',
  'C#N',
  'C(O)=O',
  'c1ccccc1',
  'NC',
  'Br',
  'N1CCOCC1',
  'OC(C)=O',
  'C(F)(F)F',
  'N(=O)=O',
  'CO',
  'NO'],
 'F': ['Cl',
  'OC',
  'C',
  'C(F)(F)F',
  'Br',
  'C#N',
  'O',
  'N(=O)=O',
  'I',
  'OC(F)(F)F',
  'N',
  'c1ccccc1',
  'CC',
  'S(C)(=O)=O',
  'N(C)C',
  'C(C)C',
  'OCC',
  'SC',
  'C(C)(C)C'],
 'N': ['O',
  'NC',
  'C',
  'OC',
  'N(C)C',
  'NC(C)=O',
  'N(=O)=O',
  'Cl',
  'F',
  'NCC',
  'N1CCOCC1',
  'N=C(N)N',
  'OCC',
  'Br',
  'C(F)(F)F',
  'C(O)=O',
  'NS(C)(=O)=O',
  'NC(C)C',
  'N1CCCC1'],
 'c1ccccc1': ['c1ccc(F)cc1',
  'c1ccc(Cl)cc1',
  'c1ccc(OC)cc1',
  'C',
  'c1ccc(C)cc1',
  'c1cccc(Cl)c1',
  'Cc1ccccc1',
  'C(C)C',
  'C1CCCCC1',
  'c1cccc(F)c1',
  'c1cccc(OC)c1',
  'c1ccccc1OC',
  'c

In [116]:
# Then perform multi-hot encoding 
mult_encod = np.zeros((len(lis_sub),len(lis_ana)))
for i in range(len(lis_sub)):
    sub = lis_sub[i] # substitution sting 
    ana = dic[sub]   # analogous
    for j in range(len(lis_ana)):
        if lis_ana[j] in ana:
            mult_encod[i][j] = 1

In [117]:
mult_encod

array([[0., 1., 1., ..., 0., 0., 0.],
       [1., 0., 1., ..., 0., 0., 0.],
       [1., 1., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [1., 1., 1., ..., 0., 1., 0.],
       [1., 1., 0., ..., 0., 0., 0.]])

In [118]:
mult_encod_frame = pd.DataFrame(mult_encod,index=lis_sub,columns=lis_ana)

In [119]:
#Saved Y dataset
mult_encod_frame.to_csv('Y_train.csv')

<b> Deep learning model </b>

In [ ]:
!pip install keras

In [120]:
X = data

In [121]:
X = X.drop('Value',axis=1)

In [122]:
#select only compound in the list_sub
p = [21*i for i in range(100)]

In [123]:
X_train = X.iloc[p,:]

In [124]:
X_train = X_train.iloc[:,1:3]

In [125]:
X_train

PCA1      PCA2
0    -0.925417 -1.440281
21   -0.922884 -1.427348
42   -0.902455 -1.390604
63   -0.933411 -1.442248
84   -1.370814  0.129142
...        ...       ...
1995 -1.322616 -0.595493
2016 -1.068004 -1.241374
2037 -1.493633  0.649715
2058 -1.299460 -0.109455
2079 -1.157299 -1.053419

[100 rows x 2 columns]

In [126]:
Y_train = mult_encod_frame

In [127]:
Y_train

C   CC   Cl   OC    F  C(F)(F)F  C(C)C  c1ccccc1  CCC   Br  ...  \
C          0.0  1.0  1.0  1.0  1.0       1.0    1.0       1.0  1.0  1.0  ...   
O          1.0  0.0  1.0  1.0  1.0       1.0    0.0       1.0  0.0  1.0  ...   
F          1.0  1.0  1.0  1.0  0.0       1.0    1.0       1.0  0.0  1.0  ...   
N          1.0  0.0  1.0  1.0  1.0       1.0    0.0       0.0  0.0  1.0  ...   
c1ccccc1   1.0  0.0  0.0  0.0  0.0       0.0    1.0       0.0  0.0  0.0  ...   
...        ...  ...  ...  ...  ...       ...    ...       ...  ...  ...  ...   
N(CC)CC    1.0  0.0  0.0  1.0  0.0       0.0    0.0       0.0  0.0  0.0  ...   
CN         1.0  0.0  0.0  1.0  0.0       0.0    0.0       0.0  0.0  0.0  ...   
c1ccccc1C  0.0  0.0  0.0  0.0  0.0       0.0    0.0       1.0  0.0  0.0  ...   
I          1.0  1.0  1.0  1.0  1.0       1.0    1.0       1.0  0.0  1.0  ...   
NCC        1.0  1.0  0.0  1.0  0.0       0.0    0.0       0.0  0.0  0.0  ...   

           CNC  CN(C)C  CNC(C)=O  C(C)N  CNC(=O)CN  CN1CCOCC1  CN(C)C(N)=N  \
C          0.0     0.0       0.0    0.0        0.0        0.0          0.0   
O          0.0     0.0       0.0    0.0        0.0        0.0          0.0   
F          0.0     0.0       0.0    0.0        0.0        0.0          0.0   
N          0.0     0.0       0.0    0.0        0.0        0.0          0.0   
c1ccccc1   0.0     0.0       0.0    0.0        0.0        0.0          0.0   
...        ...     ...       ...    ...        ...        ...          ...   
N(CC)CC    0.0     0.0       0.0    0.0        0.0        0.0          0.0   
CN         1.0     1.0       1.0    1.0        1.0        1.0          1.0   
c1ccccc1C  0.0     0.0       0.0    0.0        0.0        0.0          0.0   
I          0.0     0.0       0.0    0.0        0.0        0.0          0.0   
NCC        0.0     0.0       0.0    0.0        0.0        0.0          0.0   

           CN1CCCC1  C#C  NC1CCCC1  
C               0.0  0.0       0.0  
O               0.0  0.0       0.0  
F               0.0  0.0       0.0  
N               0.0  0.0       0.0  
c1ccccc1        0.0  0.0       0.0  
...             ...  ...       ...  
N(CC)CC         0.0  0.0       0.0  
CN              0.0  0.0       0.0  
c1ccccc1C       0.0  0.0       0.0  
I               0.0  1.0       0.0  
NCC             0.0  0.0       0.0  

[100 rows x 413 columns]

In [128]:
#How to find appropriate deep learning neural network
#How to perform the model for .....
# How to perform leave-out one cross validation 

In [129]:
from keras.models import Sequential
from keras.layers import Dense
#create model
import tensorflow as tf
from tensorflow.keras.models import Sequential
model = Sequential()

#get number of columns in training data
ncx,ncy = Y_train.shape
n_cols = X_train.shape[1]

#add model layers
model.add(Dense(10000, activation='relu', input_shape=(n_cols,)))
model.add(Dense(10000, activation='relu'))
model.add(Dense(10000, activation='relu'))
model.add(Dense(ncy,activation='sigmoid'))


In [130]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_train, Y_train, test_size=.2)

In [131]:
model.compile(optimizer=tf.keras.optimizers.Adam(0.01),loss='binary_crossentropy',metrics=['accuracy'])
model.fit(X_train,Y_train,epochs=30,batch_size=10,validation_split=0.2,verbose=0.2)


Epoch 1/30
Epoch 2/30
Epoch 3/30
Epoch 4/30
Epoch 5/30
Epoch 6/30
Epoch 7/30
Epoch 8/30
Epoch 9/30
Epoch 10/30
Epoch 11/30
Epoch 12/30
Epoch 13/30
Epoch 14/30
Epoch 15/30
Epoch 16/30
Epoch 17/30
Epoch 18/30
Epoch 19/30
Epoch 20/30
Epoch 21/30
Epoch 22/30
Epoch 23/30
Epoch 24/30
Epoch 25/30
Epoch 26/30
Epoch 27/30
Epoch 28/30
Epoch 29/30
Epoch 30/30


In [132]:
y_hat = model.predict(X_test)

1/1 [==============================] - 0s 126ms/step


In [133]:
y_hat

array([[8.4942269e-01, 1.8224184e-01, 4.3195137e-01, ..., 8.1090735e-07,
        7.1770722e-07, 8.7665524e-07],
       [7.0381188e-01, 3.5655791e-01, 4.0916616e-01, ..., 4.3088362e-06,
        4.5475840e-06, 4.5938623e-06],
       [7.5907546e-01, 4.5615798e-01, 4.6235320e-01, ..., 6.9739649e-06,
        6.3165685e-06, 7.1088280e-06],
       ...,
       [5.6876040e-01, 2.6141304e-01, 3.4142914e-01, ..., 1.0579013e-05,
        1.1164625e-05, 1.1261755e-05],
       [5.4227477e-01, 2.9965985e-01, 3.4933844e-01, ..., 2.1467171e-05,
        2.1866175e-05, 2.2278793e-05],
       [4.5728931e-01, 4.5084593e-01, 5.4353905e-01, ..., 2.1113590e-05,
        1.9709416e-05, 2.0559131e-05]], dtype=float32)

In [134]:
y_test
y_test.columns

Index(['C', 'CC', 'Cl', 'OC', 'F', 'C(F)(F)F', 'C(C)C', 'c1ccccc1', 'CCC',
       'Br',
       ...
       'CNC', 'CN(C)C', 'CNC(C)=O', 'C(C)N', 'CNC(=O)CN', 'CN1CCOCC1',
       'CN(C)C(N)=N', 'CN1CCCC1', 'C#C', 'NC1CCCC1'],
      dtype='object', length=413)

In [135]:
Result = pd.DataFrame(y_hat,index=y_test.index,columns=y_test.columns)


In [136]:
Result 

C        CC        Cl        OC         F  \
c1ccc(Cl)c(Cl)c1      0.849423  0.182242  0.431951  0.607719  0.074581   
Oc1ccccc1             0.703812  0.356558  0.409166  0.340070  0.345423   
OCc1ccccc1            0.759075  0.456158  0.462353  0.472154  0.361271   
c1ccc(S(N)(=O)=O)cc1  0.861659  0.279003  0.339805  0.527505  0.070583   
C1CCCC1               0.594285  0.255375  0.356809  0.550170  0.313297   
C(=O)OC               0.564632  0.267590  0.342147  0.630096  0.342381   
C(=O)OC(C)(C)C        0.680146  0.336641  0.365026  0.412939  0.250802   
c1ccccc1F             0.708574  0.362526  0.413954  0.349624  0.364363   
CC                    0.487287  0.427793  0.498277  0.598045  0.488048   
NCC                   0.499714  0.368373  0.439356  0.649864  0.451612   
c1cccc(OC)c1          0.757660  0.457787  0.457862  0.467171  0.364668   
OC                    0.484176  0.422859  0.497570  0.599926  0.488297   
I                     0.642464  0.335776  0.374841  0.451694  0.291145   
c1ccccc1              0.673881  0.357733  0.376190  0.400576  0.272205   
C(O)=O                0.502934  0.357251  0.426088  0.655988  0.443343   
c1ccc(C)cc1           0.702690  0.361217  0.412574  0.345373  0.354159   
C#N                   0.481516  0.431424  0.506974  0.591206  0.494347   
CC(N)=O               0.568760  0.261413  0.341429  0.623137  0.337781   
CC(C)C                0.542275  0.299660  0.349338  0.654917  0.366830   
C                     0.457289  0.450846  0.543539  0.548587  0.526053   

                      C(F)(F)F     C(C)C  c1ccccc1       CCC        Br  ...  \
c1ccc(Cl)c(Cl)c1      0.040692  0.313513  0.973857  0.158882  0.024578  ...   
Oc1ccccc1             0.171492  0.397494  0.636248  0.168817  0.093202  ...   
OCc1ccccc1            0.133671  0.392521  0.756017  0.149954  0.102635  ...   
c1ccc(S(N)(=O)=O)cc1  0.064150  0.411783  0.964828  0.251423  0.035920  ...   
C1CCCC1               0.331160  0.361086  0.682489  0.243991  0.255780  ...   
C(=O)OC               0.330569  0.396144  0.747338  0.187948  0.270883  ...   
C(=O)OC(C)(C)C        0.288786  0.410125  0.591223  0.229998  0.189138  ...   
c1ccccc1F             0.170080  0.403212  0.652357  0.163922  0.095656  ...   
CC                    0.499960  0.517153  0.830723  0.068719  0.348874  ...   
NCC                   0.431121  0.483252  0.808342  0.088417  0.321155  ...   
c1cccc(OC)c1          0.136824  0.396424  0.754522  0.151486  0.104505  ...   
OC                    0.498471  0.513209  0.833568  0.067545  0.347539  ...   
I                     0.318032  0.409837  0.608709  0.256909  0.232911  ...   
c1ccccc1              0.283477  0.420852  0.578760  0.243287  0.183690  ...   
C(O)=O                0.417285  0.476101  0.802166  0.093753  0.316029  ...   
c1ccc(C)cc1           0.175331  0.402006  0.639310  0.171352  0.098078  ...   
C#N                   0.509437  0.518102  0.837810  0.064200  0.351199  ...   
CC(N)=O               0.329296  0.387997  0.742901  0.195336  0.268044  ...   
CC(C)C                0.339916  0.432546  0.763806  0.146352  0.279923  ...   
C                     0.570165  0.534484  0.862764  0.047593  0.361550  ...   

                               CNC        CN(C)C      CNC(C)=O         C(C)N  \
c1ccc(Cl)c(Cl)c1      6.773286e-07  7.413532e-07  6.779192e-07  7.249979e-07   
Oc1ccccc1             4.289538e-06  4.248463e-06  4.604859e-06  4.396930e-06   
OCc1ccccc1            5.794660e-06  6.373545e-06  5.962408e-06  6.409452e-06   
c1ccc(S(N)(=O)=O)cc1  1.928611e-06  1.851336e-06  1.794534e-06  1.928109e-06   
C1CCCC1               1.149860e-05  1.204135e-05  1.213090e-05  1.011994e-05   
C(=O)OC               1.316030e-05  1.421688e-05  1.352771e-05  1.165044e-05   
C(=O)OC(C)(C)C        1.040971e-05  1.046337e-05  1.124702e-05  9.702211e-06   
c1ccccc1F             5.106268e-06  5.149916e-06  5.464157e-06  5.331709e-06   
CC                    3.146018e-05  3.198520e-05  3.088214e-05  3.114005e-05   
NCC           

The result will be the possibilty value for each substitute to have this analogous

This is the example coding material. 

The full research work cannot be publicied. 

In [138]:
# Save model 
model.save('tfmodel')

INFO:tensorflow:Assets written to: tfmodel\assets
